Change number parsing to treat 2. as a number

drlippman opened the pull request:
As identified in issue #124, in the current parsing `2.3`, `2` and `.2` are treated as numbers, but `2.` is treated as a number followed by an period operator.  Since trailing decimal places can be used to indicate trailing zeros are significant, this behavior can cause valid numbers to not be treated as such.

However, this also means that if a end-of-sentence period follows a number, and that period is inside the asciimath area for some reason, that period would be treated as part of the number, which may not be desirable.  But arguably this case would be caused by semantically incorrect markup, so perhaps isn't something we need to worry about.

drlippman committed:
Change number parsing to treat 2. as a number

drlippman committed:
Merge branch 'master' into feature.trailing_decimal

drlippman commented:
@jipsen, since you understand the parsing grammar much better than me, it'd be great to get your thoughts of whether this would have any unintentional side effects.

drlippman commented:
There's a significant issue with this change when `decimalsign=","`, since it would break list and matrix parsing.  Either this would have to be limited to when a period is used as decimalsign, or it would need some additional logic to determine when to include the decimalsign as part of the number.

drlippman committed:
Adjust handling for non-period decimal separators

"3,4" and ",5" will be treated as numbers,
as well as stuff like "3,/4" or "(3,)"
but "3, " won't include the comma in the number to prevent
breaking lists and matrices which rely on a space after a
comma being used to distinguish a list separator

drlippman commented:
I made a change where "3,4" and ",5" will be treated as numbers, as well as stuff like "3,/4" or "(3,)" but "3, " won't include the comma in the number to prevent breaking lists and matrices which rely on a space after a comma being used to distinguish a list separator.  When decimalsign="." it will still treat "3. " as a number. I'm not sure if this is a reasonable compromise.

christianp commented:
If the decimal separator is a comma, shouldn't the list separator be a semicolon?

drlippman commented:
A semicolon list separator would certainly be clearer, but if the system required use of a semicolon and broke display when using a comma, that'd be a major breaking change for all existing documents written based on the current behavior.  I'm not familiar with how math is written in countries that use comma as decimal separator, but I'm guessing it might not be uncommon to still use commas in lists as well.

I suppose we could add a new listSeparator config option. A conservative backwards-compatible handling could be used when listSeparator is the default comma, and a more aggressive handling that would treat "3," as a single number when listSeparator is something else.

drlippman committed:
add listseparator config option

allows setting it to something like ; when decimalsign=","
which allows for clearer parsing of matrices and lists,
without needing a space after the listseparator.  Also allows
"3," to be safely interpreted as a single number.

drlippman committed:
update asciimath-based

drlippman committed:
Merge branch 'master' into feature.trailing_decimal

drlippman committed:
Merge branch 'master' into feature.trailing_decimal

drlippman committed:
Increment version number
